docs/src/modules/utils/StyledEngineProvider.js

```diff
@@ -16,15 +16,23 @@ const cacheRtl = createCache({
 });
 
 export default function StyledEngineProvider(props) {
-  const { children, cacheLtr } = props;
+  const { children } = props;
   const theme = useTheme();
-
   const rtl = theme.direction === 'rtl';
-  const emotionCache = theme.direction === 'rtl' ? cacheRtl : cacheLtr;
 
+  let Wrapper = React.Fragment;
+  let wraperProps = {};
+
+  // Only happens client-side after the hydration
+  if (rtl) {
+    Wrapper = CacheProvider;
+    wraperProps = {
```

Comment:
> The main change is reverting https://github.com/mui/material-ui/pull/28495

This seems something that we should keep, performance should be worse if we remount each DOM node, no? Could the root of the problem be a regression in emotion?

Reply:
> Could the root of the problem be a regression in emotion?

It could be, I am trying to pinpoint to what exact change did the perf regression. I am still testing things out, there seems to be big difference between components pages vs other changes (and I remember we did optimization on the component pages, e.g. using common layout etc). Maybe there is something there too.